storage/entity.go

```diff
@@ -0,0 +1,353 @@
+		case uuid.UUID:
+			completeMap[typeKey] = OdataGUID
+			completeMap[k] = t.String()
+		case int64:
+			completeMap[typeKey] = OdataInt64
+			completeMap[k] = fmt.Sprintf("%v", v)
+		default:
+			completeMap[k] = v
+		}
+		if strings.HasSuffix(k, OdataTypeSuffix) {
+			if !(completeMap[k] == OdataBinary ||
+				completeMap[k] == OdataDateTime ||
+				completeMap[k] == OdataGUID ||
+				completeMap[k] == OdataInt64) {
+				return nil, fmt.Errorf("Odata.type annotation %v value is not valid", k)
+			}
+			valueKey := strings.TrimSuffix(k, OdataTypeSuffix)
+			if _, ok := completeMap[valueKey]; !ok {
+				return nil, fmt.Errorf("Odata.type annotation %v defined without value defined", k)
+			}
+		}
+	}
+	return json.Marshal(completeMap)
+}
+
+// UnmarshalJSON is a custom unmarshaller for entities
+func (e *Entity) UnmarshalJSON(data []byte) error {
+	errorTemplate := "Deserializing error: %v"
+
+	props := map[string]interface{}{}
+	err := json.Unmarshal(data, &props)
+	if err != nil {
+		return err
+	}
+
+	// deselialize metadata
+	props = assignFromMap(props, &e.OdataMetadata, "odata.metadata")
+	props = assignFromMap(props, &e.OdataType, "odata.type")
+	props = assignFromMap(props, &e.OdataID, "odata.id")
+	props = assignFromMap(props, &e.OdataEtag, "odata.etag")
+	props = assignFromMap(props, &e.OdataEditLink, "odata.editLink")
+	props = assignFromMap(props, &e.PartitionKey, partitionKeyNode)
+	props = assignFromMap(props, &e.RowKey, rowKeyNode)
+
+	// deserialize timestamp
+	str, ok := props["Timestamp"].(string)
+	if !ok {
+		return fmt.Errorf(errorTemplate, "Timestamp casting error")
+	}
+	t, err := time.Parse(time.RFC3339Nano, str)
+	if err != nil {
+		return fmt.Errorf(errorTemplate, err)
+	}
+	e.TimeStamp = t
+	delete(props, "Timestamp")
+	delete(props, "Timestamp@odata.type")
+
+	// deserialize entity (user defined fields)
+	for k, v := range props {
+		if strings.HasSuffix(k, OdataTypeSuffix) {
+			valueKey := strings.TrimSuffix(k, OdataTypeSuffix)
+			str, ok := props[valueKey].(string)
+			if !ok {
+				return fmt.Errorf(errorTemplate, fmt.Sprintf("%v casting error", v))
+			}
+			switch v {
+			case OdataBinary:
+				props[valueKey] = []byte(str)
+			case OdataDateTime:
+				t, err := time.Parse("2006-01-02T15:04:05Z", str)
+				if err != nil {
+					return fmt.Errorf(errorTemplate, err)
+				}
+				props[valueKey] = t
+			case OdataGUID:
+				props[valueKey] = uuid.FromStringOrNil(str)
+			case OdataInt64:
+				i, err := strconv.ParseInt(str, 10, 64)
+				if err != nil {
+					return fmt.Errorf(errorTemplate, err)
+				}
+				props[valueKey] = i
+			default:
+				return fmt.Errorf(errorTemplate, fmt.Sprintf("%v is not supported", v))
+			}
+			delete(props, k)
+		}
+	}
+
+	e.Properties = props
+	return nil
+}
+
+func getAndDelete(props map[string]interface{}, key string) (map[string]interface{}, string) {
+	str, ok := props[key].(string)
+	if ok {
+		delete(props, key)
+		return props, str
+
+	}
+	return props, ""
+}
```

Comment:
The method above needs to be repaired.

At the moment, it looks to me like getAndDelete is supposed to test to see if a key is in the map then delete it if present. Otherwise, return the same map and an empty string.
However, what it will actually do is validate that the value stored at `props[key]` is a string, then delete it, otherwise return the unaltered map and an empty string.
I wrote up a sample to demonstrate the issue here: https://play.golang.org/p/R1PsVWH_AV

The updated function could look something like this:
``` Go
func getAndDelete(props map[string]interface{}, key string) (map[string]interface{}, interface{}) {
    val, ok := props[key]
    if ok {
        delete(props, key)
        return props, val
    }
    return props, ""
}
```